pkg/kubectl/cmd/annotation.go

```diff
@@ -0,0 +1,211 @@
+
+// Update pod 'foo' only if the resource is unchanged from version 1.
+$ kubectl annotation pods foo source_id=foo-v1:c70ce5c8-3165-11e5-a151-feff819cdc9f --resource-version=1
+
+// Update pod 'foo' by removing an annotation named 'source_id' if it exists.
+// Does not require the --overwrite flag.
+$ kubectl annotation pods foo source_id-`
+)
+
+func NewCmdAnnotation(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "annotation [--overwrite] RESOURCE NAME KEY_1=VAL_1 ... KEY_N=VAL_N [--resource-version=version]",
+		Short:   "Update the annotations on a resource",
+		Long:    annotation_long,
+		Example: annotation_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunAnnotation(f, out, cmd, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().Bool("overwrite", false, "If true, allow annotations to be overwritten, otherwise reject annotation updates that overwrite existing annotations.")
+	cmd.Flags().Bool("all", false, "select all resources in the namespace of the specified resource types")
+	cmd.Flags().String("resource-version", "", "If non-empty, the annotation update will only succeed if this is the current resource-version for the object. Only valid when specifying a single resource.")
+	return cmd
+}
+
+// validateNoAnnotationOverwrites validates that when overwrite is false, to-be-updated annotations don't exist in the object annotation map (yet)
+func validateNoAnnotationOverwrites(meta *api.ObjectMeta, annotations map[string]string) error {
+	for key := range annotations {
+		if value, found := meta.Annotations[key]; found {
+			return fmt.Errorf("'%s' already has a value (%s), and --overwrite is false", key, value)
+		}
+	}
+	return nil
+}
+
+// parseAnnotations parses the annotaion arg and returns annotation maps to update and annotaions to remove
+func parseAnnotations(spec []string) (annotations map[string]string, remove []string, err error) {
+	annotations = map[string]string{}
+	for _, annotationSpec := range spec {
+		if strings.Index(annotationSpec, "=") != -1 {
+			parts := strings.SplitN(annotationSpec, "=", 2)
+			if len(parts) != 2 || len(parts[1]) == 0 {
+				return nil, nil, fmt.Errorf("invalid annotation spec: %v", annotationSpec)
+			}
+			annotations[parts[0]] = parts[1]
+		} else if strings.HasSuffix(annotationSpec, "-") {
+			remove = append(remove, annotationSpec[:len(annotationSpec)-1])
+		} else {
+			return nil, nil, fmt.Errorf("unknown annotation spec: %v", annotationSpec)
+		}
+	}
+	for _, removeAnnotation := range remove {
+		if _, found := annotations[removeAnnotation]; found {
+			return nil, nil, fmt.Errorf("can not both modify and remove an annotation in the same command")
+		}
+	}
+	return
+}
+
+func annotationFunc(obj runtime.Object, overwrite bool, resourceVersion string, annotations map[string]string, remove []string) (runtime.Object, error) {
```

Comment:
Also, doesn't this mutate the original?  I don't necessarily object, but it makes returning a `runtime.Object` seem a little deceptive.